### PR TITLE
Handlebars: improved typings

### PR DESCRIPTION
**Pull request by evil-shrike**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://handlebarsjs.com/reference.html
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

The major motivation for the update is very poor typing for helper's signature (it was just `Function`).

### CODE REVIEW

File: types/handlebars/index.d.ts [hunk 1]
```diff
@@ -1,15 +1,48 @@
...
 
 declare namespace Handlebars {
-    export function registerHelper(name: string, fn: Function, inverse?: boolean): void;
-    export function registerHelper(name: Object): void;
-    export function registerPartial(name: string, str: any): void;
+    export interface TemplateDelegate<T = any> {
+        (context: T, options?: RuntimeOptions): string;
+    }
+
+    export type Template<T = any> = TemplateDelegate<T>|string;
```
Comment (sheetalkamat) [inline types/handlebars/index.d.ts]: Whats the use of declaring type parameter T if you arent using it anywhere in the definitions..  ?

**Review by sheetalkamat — CHANGES_REQUESTED** [review]
I dont think using type parameter makes sense given you do not instantiate them anywhere in the usage.

File: types/handlebars/index.d.ts (hunk 1 above)
Comment (evil-shrike) [reply types/handlebars/index.d.ts]: @sheetalkamat 
well, honestly I don't see much sense in type argument here, but it was there before so I didn't want to break definition.
strictly speaking T does used in the definition:
```
    export interface TemplateDelegate<T = any> {
        (context: T, options?: RuntimeOptions): string;
    }
```
so you can get a typed template:
```
let tmpl = template<MyType>(source);
// ok
tmpl(new MyType());
// error
tmpl({});
```

Comment (evil-shrike) [reply types/handlebars/index.d.ts]: @sheetalkamat ping

Comment (sheetalkamat) [reply types/handlebars/index.d.ts]: what i am asking is are the usages in HelperOptions and others using these template and template delegates will always be any? Do you need type parameter in there?

Comment (evil-shrike) [reply types/handlebars/index.d.ts]: sorry, not sure I understand the point, I provided an example above where you specify a type for template and it restricts usage of that template to instances of that type. So it makes sense for TemplateDelegate. For HelperOptions and HelperDelegate it's a bit different picture. You don't use helpers directly, Handlebars calls them. So specifying a type for context has less sense but possible
```
Handlebars.registerHelper("xxx", (context, options) => {
  return "";
});
```
Are you suggesting adding a T to registerHelper?

**Pull request merged**